### PR TITLE
Convert Log Page to sortable table

### PR DESCRIPTION
**Pull request by BramGruneir**
Now with sorting!

screen shot 2015-07-29 at 16 01 26

### CODE REVIEW

File: ui/ts/models/log.ts [hunk 1]
```diff
@@ -46,17 +48,14 @@ module Models {
...
-          });
-      });
+      private _data: Utils.QueryCache<Proto.LogEntry[]>;
 
       private _url(): string {
         let url: string = "/_status/logs/";
         if (this.node() != null) {
           url += this.node();
+        } else {
+          url += "local";
```
Comment (tamird) [inline ui/ts/models/log.ts]: what's this?

Comment (BramGruneir) [inline ui/ts/models/log.ts]: if you go to https://localhost:8080/#/logs without a node id, it will show you the local logs. It used to work with no node id, but that's no longer the case since the updated .  Really it doesn't fit with the overall model of the ui but I didn't want to remove the feature yet.

File: ui/ts/models/log.ts [hunk 2]
```diff
@@ -77,13 +76,22 @@ module Models {
...
-      constructor() {
+      constructor(nodeId?: string) {
         this.level(Utils.Format.Severity(2));
         this.max(null);
         this.startTime(null);
         this.endTime(null);
         this.pattern(null);
-        this.node(null);
+        this.node(nodeId);
+        // Add the QueryCache after initialization in order to not fetch without
```
Comment (tamird) [inline ui/ts/models/log.ts]: can you explain this comment and the code movement it refers to? how could the fetch ever happen during the object's construction?

File: ui/ts/models/log.ts (hunk 1 above)
Comment (tamird) [inline ui/ts/models/log.ts]: I'm not sure I follow. This will be required soon but isn't yet?

File: ui/ts/models/log.ts [hunk 3]
```diff
@@ -77,13 +76,22 @@ module Models {
...
+        // the node id.
+        this._data = new Utils.QueryCache((): Promise<Proto.LogEntry[]> => {
+          return m.request({ url: this._url(), method: "GET", extract: nonJsonErrors })
+            .then((results: LogResponseSet) => {
+              return results.d;
+            });
+        });
+        this.allEntries = this._data.result;
```
Comment (tamird) [inline ui/ts/models/log.ts]: what's the difference between this setter syntax and the method call syntax above?

File: ui/ts/pages/log.ts [hunk 4]
```diff
@@ -119,11 +134,15 @@ module AdminViews {
...
-        if (entries.result() != null) {
-          for (let i: number = 0; i < entries.result().length; i++) {
-            rows.push(_EntryRow(entries.result()[i], i));
-          }
+        let comparisonData: Table.TableData<LogEntry> = {
+          columns: ctrl.columns,
+          rows: entries.allEntries
+        };
+        let count: number;
+        if (entries.allEntries()) {
```
Comment (tamird) [inline ui/ts/pages/log.ts]: how can `allEntries` be null? (assuming that's what this checks)

File: ui/ts/models/log.ts (hunk 1 above)
Comment (BramGruneir) [inline ui/ts/models/log.ts]: Ah, I missed a word in the explanation.  Let me clarify.
In the old status server code, going to /_status/logs/ (without a node id at the end) would return the local logs.  This is no longer the case, you have to explicitly add the id or "local" (e.g. /_status/logs/1, /_status/logs/local).  This change deals with this.

However, getting the local logs doesn't really work with the current UI design, so I plan to remove the "log" link up top that leads to this local log page in a subsequent PR.

Comment (tamird) [inline ui/ts/models/log.ts]: Ah okay. Can this change get its own commit, at least? It seem to be unrelated to the rest of the change here.

File: ui/ts/models/log.ts (hunk 2 above)
Comment (BramGruneir) [inline ui/ts/models/log.ts]: When the queryCache was up top, it immediately starts a query, which happens during the construction of the object, not after the constructor runs.  So the first query it sends is to a "null" nodeID ("local").  Which then gets refreshed after the constructor is called.  Moving the code here stops that first call from happening prematurely.
I'm working on cleaner solution right now, but doing it this way is perfectly acceptable.

Comment (tamird) [inline ui/ts/models/log.ts]: Ah, it wasn't clear that the query cache would eagerly dispatch a query on construction.

File: ui/ts/models/log.ts (hunk 3 above)
Comment (BramGruneir) [inline ui/ts/models/log.ts]: allEntries is just a public alias for _data.result.  It's not a setter.

File: ui/ts/pages/log.ts (hunk 4 above)
Comment (BramGruneir) [inline ui/ts/pages/log.ts]: This does indeed check for null.
Until after the first query returns, the results are null.

File: ui/ts/models/log.ts (hunk 3 above)
Comment (tamird) [inline ui/ts/models/log.ts]: `foo.bar = ` is not a setter?

File: ui/ts/models/log.ts (hunk 1 above)
Comment (BramGruneir) [inline ui/ts/models/log.ts]: It's a tiny bug fix.  
But you're right, it is unrelated.  If it was more substantial than a very simple fix, I'd agree.

File: ui/ts/models/log.ts (hunk 2 above)
Comment (BramGruneir) [inline ui/ts/models/log.ts]: For the most part, we actually want it to be eager.  For node/store status, ts values, and more, it doesn't require an id in the url.
I'm adding the ability to have it wait for an init call in a subsequent pr right now.

Comment (BramGruneir) [inline ui/ts/models/log.ts]: But I'll adjust the comment to be more direct.

File: ui/ts/models/log.ts (hunk 1 above)
Comment (tamird) [inline ui/ts/models/log.ts]: `git reset head^ --patch -- ui/ts/models/log.ts`; select this bit; `git commit --amend`; `git add ui/ts/models/log.ts`; `git commit -m 'tiny bug fix'`

File: ui/ts/pages/log.ts (hunk 4 above)
Comment (tamird) [inline ui/ts/pages/log.ts]: oh, `allEntries` is not null, but it's being called. OK.

File: ui/ts/models/log.ts (hunk 3 above)
Comment (mrtracy) [inline ui/ts/models/log.ts]: `this._data.result` contains a property, which is a getter/setter function.  This line is assigning that function to `this.allEntries`.

Comment (BramGruneir) [inline ui/ts/models/log.ts]: It's not a _mithril.MithrilProperty like the rest of the values here.

Comment (tamird) [inline ui/ts/models/log.ts]: with you so far. what does `this.pattern(null);` above do? is it a setter method?

Comment (tamird) [inline ui/ts/models/log.ts]: ah, ok, it's a special thing. thanks!

File: ui/ts/pages/log.ts [hunk 5]
```diff
@@ -48,49 +106,6 @@ module AdminViews {
         return new Controller(nodeId);
       };
 
-      // TODO(bram): Move these into css classes.
-      const _tableStyle: string = "border-collapse:collapse; border - spacing:0; border - color:#ccc";
```
Comment (mrtracy) [inline ui/ts/pages/log.ts]: :lgtm

File: ui/ts/pages/log.ts [hunk 6]
```diff
@@ -18,14 +19,71 @@ module AdminViews {
...
+            title: "Severity",
+            view: (entry: LogEntry): string => Utils.Format.Severity(entry.severity)
+          },
+          {
+            title: "Message",
+            view: (entry: LogEntry): string => Utils.Format.LogEntryMessage(entry)
+          },
+          {
+            title: "Node",
+            view: (entry: LogEntry): string => Utils.Convert.NumberToString(entry.node_id),
```
Comment (mrtracy) [inline ui/ts/pages/log.ts]: Use builtin function `String()`, it works with everything and is null safe.

File: ui/ts/models/log.ts (hunk 1 above)
Comment (BramGruneir) [inline ui/ts/models/log.ts]: Done.

File: ui/ts/pages/log.ts (hunk 6 above)
Comment (BramGruneir) [inline ui/ts/pages/log.ts]: I had that originally.  Let me try it again.

Comment (BramGruneir) [inline ui/ts/pages/log.ts]: Fixed and removed the fully convert function.

**Comment by mrtracy** [comment]
LGTM

**Comment by mrtracy** [comment]
Still LGTM
